Add auto configuration script

schrej opened the pull request:
A script that allows to fetch the configuration from the panel.

Usage: `npm run configure -- -h`, the double `--` is required between arguments.
If required parameters are not provided via command line the script is interactive and asks for them.

the `-o` flag will force overwrite, without it the script will ask if it should overwrite.

There is some room for code improvements. Some stuff can be simplified probably, just created this for first looks at it.

Pterodactyl/Panel#242 & Pterodactyl/Panel#218

schrej committed:
add configure script
with working command line params and interactive prompt
add yargs and inquirer npm packages

schrej committed:
config script: require the panel url

schrej committed:
add core functionality to configure script
there is room for imrovements

schrej commented on scripts/configure.js:
```diff
@@ -0,0 +1,151 @@
+};
+
+// Validate and filter the provided command line params
+function validateParams() {
```
The params should be validated with yargs as it's possible i think. Will change that

schrej committed:
configure scripts now exists on parameter errors
some refactoring and comments

DaneEveritt commented on package.json:
```diff
@@ -52,6 +53,7 @@
     "fs-extra": "1.0.0",
     "gamedig": "0.2.25",
     "ini": "1.3.4",
+    "inquirer": "^2.0.0",
```
This should be set to `2.0.0` so its always a known version.

DaneEveritt commented on package.json:
```diff
@@ -67,6 +69,8 @@
-    "socketio-file-upload": "0.5.0"
+    "socketio-file-upload": "0.5.0",
+    "winston": "^2.3.0",
+    "yargs": "^6.6.0"
```
Same with these two. Set to hard version.

DaneEveritt commented on scripts/configure.js:
```diff
@@ -0,0 +1,166 @@
+const _ = require('lodash');
```
Need a copyright header in this file, as well as `'use strict;'`.

DaneEveritt commented on scripts/configure.js:
```diff
@@ -0,0 +1,166 @@
+const _ = require('lodash');
+const inquirer = require('inquirer');
+const winston = require('winston');
```
All constants should be capitalized according to ESLint rules.

schrej replied:
My es-lint doesn't say so

DaneEveritt commented on scripts/configure.js:
```diff
@@ -0,0 +1,166 @@
+const fs = require('fs');
+const path = require('path');
+
+const configFilePath = path.resolve(__dirname, '../config/core.json');
```
You can also use `rfr` here if you'd like, but thats up to you.

DaneEveritt commented on scripts/configure.js:
```diff
@@ -0,0 +1,166 @@
+                return 'A configuration token is required.';
+            }
+            if (value.length !== 32) {
+                return 'The token provided is invalid. It must be 32 characters long.';
```
Its generally better to make this `if (value.length === 32) return true;`. This will prevent an error or some freak thing from skipping this and returning true below. I try to make things in the daemon return false by default, and only return true if they match specific rules.

DaneEveritt commented on scripts/configure.js:
```diff
@@ -0,0 +1,166 @@
+                        log.info('Configuration written successfully.');
+                    }
+                });
+            } else if (response.statusCode === 403) {
```
What happens if its not a `200` or a `403` code?

schrej replied:
It should not be happening but I'll add a else

DaneEveritt commented:
I'd also suggest changing `npm test` to cover the scripts directory too, or move the scripts folder into `src`.

schrej committed:
switch to fixed dependency versions
add scripts/* to npm test

schrej committed:
add license and use strict
capitalize dependencies
resolve config path with rfr instead of file
handle unexpected http errors on config fetch
safer token length validation
